Make cantherm more flexible

goldmanm opened the pull request:
<!--
Thanks for contributing a pull request! Please try to provide as much detail as possible to help the reviewer understand your work.
You can also add the appropriate labels to describe the topic of the pull request and the type of changes you're making.
-->

### Motivation or Problem
Cantherm's power is not being fully utilized. For example, it is integrated into RMG, but lacks the ability to use estimation methods in RMG. Many of the hindrances are just hardcoded extraneous print methods, which cause the program to fail even when it is working properly.

This PR allows cantherm to do much more than it was originally written for. For example, with this PR you can run a pressure dependent job using the inverse laplace transform for barrierless reactions... or you can input the NASA polynomials instead of a quantum file if the quantum job is not available. 

### Description of Changes
Commits in this pull request mostly fall into a few categories:

1. They prevent the job from dying due to hardcoded extraneous functionality (like the printing of a table of rates into the comment of the output.py file, which breaks if the temperature range you specify for the job is not within the bounds hardcoded by the print statement)
2. They catch errors earlier, so you don't have to wonder why your output is all NaN, since the cause appears earlier
3. They make cantherm's behavior more intuitive. For example, if the user inputs a species with a specific ZPE-corrected E0, cantherm will use that value instead of trying to re-estimate it.

A few other commits in this PR involve allowing molecule objects to be properly accessed in data-tree structures, and refactoring methods in Species to be more modular.

### Testing
unittests. and will look at rmg-tests

### Reviewer Tips
Make sure commits make sense and run cantherm.
<!--
Checklist before submission:
 - Have you added appropriate unit tests?
 - Have you checked that all unit tests pass?
 - Is your code commented and understandable?
 - Have you updated related documentation?
 - Are the commits logically organized and informative?
 - Is your branch up to date with master?
-->

mjohnson541 commented:
The functionality for cantherm to use rate estimates from RMG and take direct input (rate or nasa polynomial) was already added in #1393.

alongd commented on rmgpy/species.py:
```diff
@@ -651,10 +656,31 @@ def generateStatMech(self):
+                # set Cp0 and CpInf
+                from rmgpy.data.thermo import findCp0andCpInf
+                findCp0andCpInf(self, self.thermo)
+            self.conformer.E0 = self.getThermoData().toWilhoit().E0
```
Why is `toWilhoit()` needed here, but not in the lines above?

alongd commented:
@goldmanm, could you take a look at the comments here? I think we should aim to merge this one before the renaming PR, #1422, to save this branch from conflicts

goldmanm commented on rmgpy/pdep/configuration.pyx:
```diff
@@ -234,7 +236,7 @@ cdef class Configuration:
             modes.extend(species.conformer.getActiveModes(activeKRotor=activeKRotor, activeJRotor=activeJRotor))
         
-        if rmgmode:
+        if rmgmode or len(modes) == 0:
```
Adding a print statement right before that shows that no `IdealGasTranslation` is used.

```
[HarmonicOscillator(frequencies=([2750,2807.14,2864.29,2921.43,2978.57,3035.71,3092.86,3150,900,928.571,957.143,985.714,1014.29,1042.86,1071.43,1100,2750,2800,2850,1350,1500,750,1050,1375,1000,300,800,800,800,800,800,800,800,800,800,800,800,800,800,800,800,1600,1600,1600,1600,1600,1600,1600,1600,1600,1600,1600,1600,1600,1600,1600],'cm^-1')), HinderedRotor(inertia=(0.156089,'amu*angstrom^2'), symmetry=1, barrier=(3.5888,'kJ/mol'), semiclassical=False)]

```

goldmanm replied:
I added this specifically because it caused a problem when the user provides no frequency information.

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -218,6 +219,17 @@ def species(label, *args, **kwargs):
+        if structure and not molecularWeight:
+            spec.molecularWeight = (spec.molecule[0].getMolecularWeight(), "g/mol")
+
+        if not energyTransferModel:
```
It definitely could use improvements

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -472,6 +497,29 @@ def adjacencyList(adj):
 def InChI(inchi):
     return Molecule().fromInChI(inchi)
 
+def loadNecessaryDatabases():
```
That database loading is only specified if the user places the 'database' block in their input file (which might also load many unnecessary functionality, like group additivity, kinetic trees, etc. which are not needed when the user provides their own best guess)

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -218,6 +219,17 @@ def species(label, *args, **kwargs):
+            spec.generateStatMech()
+
+        if structure and not molecularWeight:
+            spec.molecularWeight = (spec.molecule[0].getMolecularWeight(), "g/mol")
```
removed that and cherry-picked 70d146b

goldmanm commented on rmgpy/cantherm/kinetics.py:
```diff
@@ -46,12 +46,17 @@
 import rmgpy.constants as constants
 from rmgpy.molecule.draw import MoleculeDrawer, createNewSurface
 
+from rmgpy.species import SpeciesError
```
fixed

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -234,7 +246,7 @@ def transitionState(label, *args, **kwargs):
         job = StatMechJob(species=ts, path=path)
         jobList.append(job)
     
-    elif len(args) == 0 and len(kwargs) > 0:
```
If no QM or direct input is given, I think the proper behavior should be to throw an error (since the user is not giving cantherm enough information).

alongd replied:
Is there a reason the condition here differs from the respective condition is `species`?
In other words, shouldn't this be `if len(kwargs) > 0` instead of `elif len(args) == 0`?
Also, would you like to add an InputError like the one you added for species for `len(args) > 1`?

goldmanm commented on rmgpy/cantherm/statmech.py:
```diff
@@ -277,8 +277,14 @@ def load(self):
-            energyLog = None; E0 = energy
+            energyLog = None; E0 = energy; E0_withZPE = None
+        elif isinstance(energy, tuple) and len(energy) == 2:
+            # this is likely meant to be a quantity object with ZPE already accounted for
```
I added your suggestion to this code, and wrote it into the documentation.

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -333,8 +347,19 @@ def network(label, isomers=None, reactants=None, products=None, pathReactions=No
+            product_is_reactant  = any([frozenset(rxn.products ) == frozenset(reactant_pair) for reactant_pair in reactants])
+            if reactant_is_isomer or reactant_is_reactant or product_is_isomer or product_is_reactant:
+                pathReactions.append(rxn)
+        logging.info('Path reactions {} were found for network {}'.format([rxn.label for rxn in pathReactions], label))
```
yup

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -205,7 +206,7 @@ def species(label, *args, **kwargs):
         spec.reactive = reactive
         
-        if spec.thermo is None and spec.conformer.E0 is None:
+        if len(args) != 1 and spec.thermo is None and spec.conformer.E0 is None:
```
fixed these and also added check that throws error if args > 1 (since it is not proper input).

goldmanm commented on rmgpy/cantherm/pdep.py:
```diff
@@ -235,6 +235,16 @@ def copy(self):
+        
+        for config in self.network.isomers + self.network.reactants + self.network.products:
+            for spec in config.species:
+                assert spec.conformer.E0 is not None, 'species {0} is missing energy for its conformer'.format(spec.label)
```
yup!

goldmanm commented on rmgpy/pdep/msc.pyx:
```diff
@@ -67,6 +67,9 @@ cpdef applyModifiedStrongCollisionMethod(network, str efficiencyModel='default')
     Ngrains = network.Ngrains
     NJ = network.NJ
     
+    assert not numpy.isnan(densStates.sum()), 'Network {0} has NaN in the density of states.'\
```
yup!

goldmanm commented on rmgpy/data/base.py:
```diff
@@ -932,6 +932,8 @@ def matchNodeToStructure(self, node, structure, atoms, strict=False):
         if isinstance(group, LogicNode):
             return group.matchToStructure(self, structure, atoms, strict)
+        elif isinstance(group, Molecule):
+            return structure.isIsomorphic(group)
```
Maybe this should be a separate PR

goldmanm replied:
I'll spinoff a PR from this

goldmanm commented on rmgpy/data/statmech.py:
```diff
@@ -455,6 +458,7 @@ def __init__(self):
             'HinderedRotor': HinderedRotor,
             'IdealGasTranslation': IdealGasTranslation,
             'GroupFrequencies': GroupFrequencies,
+            'Conformer': Conformer,
```
I believe so. But I am not completely sure.

goldmanm replied:
I'll remove it to see.

goldmanm replied:
Nope. Removed it.

goldmanm commented on rmgpy/cantherm/pdep.py:
```diff
@@ -235,6 +235,16 @@ def copy(self):
+        for reaction in self.network.pathReactions:
+            transitionState = reaction.transitionState
+            if transitionState.conformer and transitionState.conformer.E0 is None:
+                transitionState.conformer.E0 = (sum([spec.conformer.E0.value_si for spec in reaction.reactants]) + reaction.kinetics.Ea.value_si,"J/mol")
```
Adding logging statement

alongd replied:
Could you add the reaction string to the logging msg?

goldmanm commented on rmgpy/cantherm/kinetics.py:
```diff
@@ -145,6 +150,11 @@ def generateKinetics(self,Tlist=None):
         """
+
+        if isinstance(self.reaction.kinetics, Arrhenius):
+            self.usedTST = False
```
good idea.

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -333,8 +347,19 @@ def network(label, isomers=None, reactants=None, products=None, pathReactions=No
+        # Only add reactions that match reactants and/or isomers
+        pathReactions = []
+        for rxn in reactionDict.values():
+            if len(rxn.reactants) > 1 and len(rxn.products) > 1:
```
sounds good.

goldmanm commented on rmgpy/chemkin.pyx:
```diff
@@ -1558,7 +1558,9 @@ def writeReactionString(reaction, javaLibrary = False):
-    file. Use verbose = True to turn on comments.  Use javaLibrary = True in order to 
+    file. Use `verbose = True` to turn on kinetics comments.
+    Use `commented = True` to comments out the entire reaction.
+    Use javaLibrary = True in order to
```
fixed

alongd commented:
Could you take a look at the Codacy report, and see what could be fixed?

alongd commented:
@goldmanm, could you take a look at the tests on this branch? Please also rebase. I'd like to get this PR in soon.

@mjohnson541, could you take a look at the database loading related code? Let us know if you have comments.

goldmanm commented on rmgpy/pdep/network.py:
```diff
@@ -643,7 +643,8 @@ def calculateMicrocanonicalRates(self):
                         logging.info(mol.toAdjacencyList())
                         logging.info('reactive = {0}\n'.format(mol.reactive))
-                raise NetworkError('Unexpected type of path reaction "{0}"'.format(rxn))
+                logging.info('Path reaction {0} not found in reaction network {1}'.format(rxn,self.label))
```
Yes. It is also beneficial to run multiple things at a time to reduce setup and shutdown issues.

goldmanm commented on rmgpy/data/kinetics/family.py:
```diff
@@ -1518,7 +1518,7 @@ def isMoleculeForbidden(self, molecule):
 
         # check RMG globally forbidden structures
-        if forbidden_structures.isMoleculeForbidden(molecule):
+        if forbidden_structures is not None and forbidden_structures.isMoleculeForbidden(molecule):
```
Thanks for the forewarning. It was removed when rebasing.

goldmanm commented:
@alongd @mjohnson541, all the tests passed except the two codecov. I'm updating the branch now.

mliu49 commented:
Is this ready to be merged? I am planning to merge #1478, but I don't want to make you have to rebase this.

goldmanm commented on rmgpy/cantherm/kinetics.py:
```diff
@@ -211,13 +213,22 @@ def save(self, outputFile):
+            try:
+                k = reaction.calculateTSTRateCoefficient(T) * factor
+                kappa = k / k0
+            except SpeciesError:
```
Good idea. Added this

goldmanm commented on rmgpy/cantherm/statmech.py:
```diff
@@ -272,15 +272,32 @@ def load(self):
+                energyLog = None; E0 = None
+            else:
+                raise InputError('The third argument for E0 energy value should'\
+                                 'be E0 (for energy w/o ZPE) or E0-ZPE. Value'\
```
added

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -165,6 +165,10 @@ def species(label, *args, **kwargs):
+    elif len(args) > 1:
+        raise InputError('species {0} can only have one non-keyword argument '
+                         'which should be the path to quantum file.'
+                         ''.format(spec.label))
```
added

goldmanm commented on rmgpy/pdep/collision.pyx:
```diff
@@ -101,8 +101,6 @@ cdef class SingleExponentialDown:
             except quantity.QuantityError:
                 self._alpha0 = quantity.Energy(value)
 
-            self._alpha0 = quantity.Energy(value)
```
This extra line was left over from some improper merging and is a bug on it's own write. If there is no error in the above try statement, it will overwrite the proper method of setting `_alpha0`.

It's in a commit about errors since it improperly is handling the QuantityError.

goldmanm commented on rmgpy/data/kinetics/family.py:
```diff
@@ -1514,7 +1514,6 @@ def isMoleculeForbidden(self, molecule):
         if self.forbidden is not None and self.forbidden.isMoleculeForbidden(molecule):
             return True
 
-
```
I think this was fixed in another PR which was merged previously. removing commit

goldmanm commented on rmgpy/cantherm/statmech.py:
```diff
@@ -362,29 +379,30 @@ def load(self):
-
+        if E0_withZPE is None:
+            # The E0 that is read from the log file is without the ZPE and corresponds to E_elec
+            if E0 is None:
```
travis no longer failing

goldmanm commented on rmgpy/cantherm/input.py:
```diff
@@ -165,6 +165,10 @@ def species(label, *args, **kwargs):
         job = StatMechJob(species=spec, path=path)
         logging.debug('Added species {0} to a stat mech job.'.format(label))
         jobList.append(job)
+    elif len(args) > 1:
```
If you see how the header is written, `args` excludes the species label which is always required.
The code you showed will still run properly since there is only one additional non-keyword argument.

I am clarifying the error message so this is not confusing to the user.

alongd commented:
Thanks for making all the changes!
The cantherm/networks/acetyl+O2 example fails on this branch, could you take a look?

alongd commented:
@goldmanm, Could you rebase and add the fix for the cantherm/networks/acetyl+O2 example we discussed offline? If all looks good I'd like to get this PR in so we can continue to two additional Cantherm/Arkane PRs. Thanks!

goldmanm commented:
@alongd, Thanks for the message. I added the fix, and thought I pushed it, but somehow didn't get to that point. I updated the documentation and example files too. I tried it on the example files, which worked as expected.
